File: 001_Python/20211005_1442_HackerrankCamelCase4Hard/came_case_4_v01.py
# Enter your code here. Read input from STDIN. Print output to STDOUT
import sys

OPERATION = 0
TYPE = 1
NAME = 2
SPLIT = 'S'
COMBINE = 'C'
METHOD = 'M'
CLASS = 'C'
VARIABLE = 'V'

output = []
for line in sys.stdin:
    output.clear()
    s = line.rstrip().split(';')
    #print(f's = {s}')
    if s[OPERATION] == SPLIT:
        #print(f'Operation = Split of name {s[NAME]}')
        if s[TYPE] == METHOD:
            # remove ()
            s[NAME] = s[NAME][: len(s[NAME]) - 2]
        # insert Spaces before each Uppercases and lower uppercases
        for c in s[NAME]:
            if c.isupper():
                output.append(' ')
                output.append(c.lower())
            else:
             output.append(c)
            #print(f'loop for : print output : {output}')
    if s[OPERATION] == COMBINE:
        # cases Method, Class, Variables
        #1st letter if Class --> upperCase
        # 1st letter after a space --> remove space, uppercase next letter
        # If method, add () at the end
        next_cap = False
        for i in range(len(s[NAME])):
            if i == 0 and s[TYPE] == CLASS:
                output.append(s[NAME][0].upper())  
            elif s[NAME][i] == ' ' and i < len(s[NAME]) -1:
                next_cap = True
            else:
                if next_cap == True:
                    output.append(s[NAME][i].upper())
                    next_cap = False
                    pass
                else:
                    output.append(s[NAME][i])
        if s[TYPE] == METHOD:
            output.append('(')
            output.append(')')

    print(''.join(output).lstrip(), file = sys.stdout)

File: 003_C++/20211005_1442_HackerrankCamelCase4Hard/came_case_4_v01.cpp
#include <cmath>
#include <cstdio>
#include <vector>
#include <iostream>
#include <algorithm>
using namespace std;

#define OPERATION 0
#define TYPE 2
#define NAME_OFFSET 4
#define SPLIT 'S'
#define COMBINE 'C'
#define METHOD 'M'
#define CLASS 'C'
#define VARIABLE 'V'

int main() {
    /* Enter your code here. Read input from STDIN. Print output to STDOUT */   
    string line, output("");
    bool flag_space = false;
    while(getline(cin,line)){
        output = "";
        if(line[OPERATION] == SPLIT){
            if(line[TYPE] == METHOD){
                // will have to remove the ending ()
                line.pop_back();
                line.pop_back();
                cout << "method line : " << line << endl;
            }             
            for(size_t i=NAME_OFFSET; i<line.size();i++){
                if(isupper(line[i])){
                    if(i != NAME_OFFSET){
                        output.push_back(' ');
                    }
                    output.push_back(tolower(line[i]));
                }
                else{
                    output.push_back(line[i]);    
                }
            }
        }
        if(line[OPERATION] == COMBINE){
            // class --> 1st Char uppercase
            // method --> ending with ()
            // Space --> skip and next char should be Uppercase
            for(size_t i=NAME_OFFSET; i<line.size();i++){
                if((i == NAME_OFFSET) && (line[TYPE] == CLASS)){
                    output.push_back(toupper(line[i]));
                }
                else if(line[i] == ' '){
                    flag_space = true;
                }
                else if(line[i] == '\r'){
                }
                else{
                    if(flag_space){
                        output.push_back(toupper(line[i]));
                        flag_space = false;
                    }
                    else{
                        output.push_back(line[i]);
                    }
                }
            }
            if(line[TYPE] == METHOD){
                output.append("()");
            }
        }
        cout << output << endl;
    }
    return 0;
}
